### PR TITLE
adding loss normalization options to rnnt joint 

### PR DESCRIPTION
**Pull request by bmwshop**
# What does this PR do ?

Adds "mean" and "mean_volume" loss normalization options to RNNT Joint

**Collection**: 
ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
In the config file, set `rnnt_reduction` to `mean`, `mean_batch` (same as the default today), or `mean_volume` in the `model` section

```python
# Add a code snippet demonstrating how to use this 
rnnt_reduction: mean_batch
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by titu1994 — COMMENTED** [review]
This should not go into joint but the RNNT loss should handle this.

File: nemo/collections/asr/modules/rnnt.py [hunk 1]
```diff
@@ -1356,7 +1359,15 @@ def forward(
             # Collect sub batch loss results
             if losses is not None:
                 losses = torch.cat(losses, 0)
-                losses = losses.mean()  # global batch size average
+                target_lengths = torch.cat(target_lengths, 0)
```
Comment (titu1994) [inline nemo/collections/asr/modules/rnnt.py:1362]: Remove all this. The RNNT loss calculates the average loss per minibatch provided to it, this mean is just to average over sub-batches.

File: nemo/collections/asr/losses/rnnt.py [hunk 2]
```diff
@@ -281,19 +281,21 @@ def forward(self, log_probs, targets, input_lengths, target_lengths):
...
 
         # Loss reduction can be dynamic, so reset it after call
-        if self.reduction != 'mean_batch':
+        if self.reduction in ['mean', 'sum']:
             self._loss.reduction = 'none'
 
-        # Loss reduction only for mean_batch mode
+        # Loss reduction only for mean_batch and mean_volume modes
         if self.reduction == 'mean_batch':
             loss = torch.mean(loss)
```
Comment (sam1373) [inline nemo/collections/asr/losses/rnnt.py:296]: What is the difference between "mean" and "mean_batch" for this loss? In the underlying loss mean reduction seems to already be described as "the output losses will be divided by the target lengths and then the mean over the batch is taken"

File: nemo/collections/asr/models/rnnt_models.py [hunk 3]
```diff
@@ -72,7 +72,10 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         loss_name, loss_kwargs = self.extract_rnnt_loss_cfg(self.cfg.get("loss", None))
 
         self.loss = RNNTLoss(
-            num_classes=self.joint.num_classes_with_blank - 1, loss_name=loss_name, loss_kwargs=loss_kwargs
+            num_classes=self.joint.num_classes_with_blank - 1,
+            loss_name=loss_name,
+            loss_kwargs=loss_kwargs,
+            reduction=self.cfg.get("rnnt_reduction", "mean_batch"),
```
Comment (sam1373) [inline nemo/collections/asr/models/rnnt_models.py:78]: probably should be in the loss config?

File: nemo/collections/asr/modules/rnnt.py [hunk 4]
```diff
@@ -1219,6 +1219,7 @@ def __init__(
 
         # Optional arguments
         dropout = jointnet.get('dropout', 0.0)
+        # self.reduction = jointnet.get('reduction', 'mean_batch')
```
Comment (sam1373) [inline nemo/collections/asr/modules/rnnt.py:1222]: remove

File: nemo/collections/asr/losses/rnnt.py [hunk 5]
```diff
@@ -281,19 +281,21 @@ def forward(self, log_probs, targets, input_lengths, target_lengths):
...
+        elif self.reduction == 'mean_volume':
+            loss = torch.sum(loss) / torch.sum(target_lengths)
```
Comment (sam1373) [inline nemo/collections/asr/losses/rnnt.py:298]: Should the underlying loss reduction be disabled for this case (seems to be mean by default)? Also, would this work properly with the fused batches?

File: nemo/collections/asr/losses/rnnt.py (hunk 2 above)
Comment (bmwshop) [reply nemo/collections/asr/losses/rnnt.py:296]: "mean_batch" - our current default - does not divide the losses by target lengths...

File: nemo/collections/asr/models/rnnt_models.py (hunk 3 above)
Comment (bmwshop) [reply nemo/collections/asr/models/rnnt_models.py:78]: This is meant to be similar to the implementation in the [CTC models](https://github.com/NVIDIA/NeMo/blob/d19146cd5cad07151f94820ff92392a29c3225f6/examples/asr/conf/conformer/conformer_ctc_bpe.yaml#L32)

File: nemo/collections/asr/losses/rnnt.py (hunk 5 above)
Comment (bmwshop) [reply nemo/collections/asr/losses/rnnt.py:298]: To the best of my understanding, the RNNT Joint [does disable ](https://github.com/NVIDIA/NeMo/blob/61ddd23eba8d88cc08d2f2f7909b75f5333cd4d1/nemo/collections/asr/modules/rnnt.py#L1330)the underlying loss reduction so that it can do its own calculation

File: nemo/collections/asr/modules/rnnt.py (hunk 1 above)
Comment (bmwshop) [reply nemo/collections/asr/modules/rnnt.py:1362]: To the best of my understanding, the RNNT Joint [does disable ](https://github.com/NVIDIA/NeMo/blob/61ddd23eba8d88cc08d2f2f7909b75f5333cd4d1/nemo/collections/asr/modules/rnnt.py#L1330)the underlying loss reduction so that it can do its own calculation

**Comment by okuchaiev** [comment]
is this necessary for r1.12 ?

File: nemo/collections/asr/modules/rnnt.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/modules/rnnt.py:1361]: No it's not. It calls the RNNTLoss which calls NumbaRNNTLoss. Those two are involved for the reductions as you have implemented in this pr. The average step in the joint is only to average these subbatch losses which have already been r reduced by the loss already.
So this file needs to be reverted

Comment (titu1994) [reply nemo/collections/asr/modules/rnnt.py:1361]: I should be clearer - the reductions we define by mean batch and mean volume utilise self.loss._reduction and not self.lpss.reduction. This arg is used only for the sum and mean types of reduction, which we remove cause it would do incorrect mean over sub-batches due to super call. 

So RNNT loss has no need to control the reductoon RNNTLoss handles it, joint just averaged losses over sub-batches

Comment (bmwshop) [reply nemo/collections/asr/modules/rnnt.py:1361]: I just took another pass at the code, with debug statements. I see that self.loss._reduction in RNNTLossNumba gets set to none.

**Review by titu1994 — APPROVED** [review]
After discussion, seems there was a silent bug for all cases other than "mean_batch" inside the fused joint. This PR fixes the issue. Since the default for RNNT was mean_batch, most model training should not be affected.

File: nemo/collections/asr/modules/rnnt.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/modules/rnnt.py:1361]: Seems to be silent bug for all cases != mean_batch. Good fix !

File: nemo/collections/asr/modules/rnnt.py [hunk 6]
```diff
@@ -1353,12 +1355,8 @@ def forward(
...
                 del sub_enc, sub_transcripts, sub_enc_lens, sub_transcript_lens
 
-            # Collect sub batch loss results
-            if losses is not None:
-                losses = torch.cat(losses, 0)
-                losses = losses.mean()  # global batch size average
-            else:
-                losses = None
+            # Reduce over sub batches
+            losses = self.loss.reduce(losses, target_lengths)
```
Comment (titu1994) [inline nemo/collections/asr/modules/rnnt.py:1359]: reduce only when not None. Keep the else block.

File: nemo/collections/asr/losses/rnnt.py [hunk 7]
```diff
@@ -243,13 +247,32 @@ def __init__(self, num_classes, reduction: str = 'mean_batch', loss_name: str =
...
-            raise ValueError('`reduction` must be one of [mean, sum, mean_batch]')
+        if reduction not in [None, 'mean', 'sum', 'mean_batch', 'mean_volume']:
+            raise ValueError('`reduction` must be one of [mean, sum, mean_batch, mean_volume]')
 
         self._blank = num_classes
         self.reduction = reduction
         self._loss = resolve_rnnt_loss(loss_name, blank_idx=self._blank, loss_kwargs=loss_kwargs)
 
+    def reduce(self, losses, target_lengths):
+        if losses is None:
```
Comment (titu1994) [inline nemo/collections/asr/losses/rnnt.py:258]: Assume losses is not None, not this classes responsibility to do this check.

File: nemo/collections/asr/losses/rnnt.py [hunk 8]
```diff
@@ -243,13 +247,32 @@ def __init__(self, num_classes, reduction: str = 'mean_batch', loss_name: str =
...
-        if reduction not in [None, 'mean', 'sum', 'mean_batch']:
-            raise ValueError('`reduction` must be one of [mean, sum, mean_batch]')
+        if reduction not in [None, 'mean', 'sum', 'mean_batch', 'mean_volume']:
+            raise ValueError('`reduction` must be one of [mean, sum, mean_batch, mean_volume]')
 
         self._blank = num_classes
         self.reduction = reduction
         self._loss = resolve_rnnt_loss(loss_name, blank_idx=self._blank, loss_kwargs=loss_kwargs)
 
+    def reduce(self, losses, target_lengths):
```
Comment (titu1994) [inline nemo/collections/asr/losses/rnnt.py:257]: Add types to this function signature

**Review by titu1994 — APPROVED** [review]
